### ISSUE TITLE
Panic with zstd dictionary compression in blob files

### ISSUE THREAD

**Issue by plenkl** [tikv/titan#220]
I have tried to enable zstd dictionary compression for the titan blob files in tikv. Unfortunately, it does not seem to work.

(The code for exposing the configuration parameters can be found here https://github.com/tikv/tikv/compare/master...plenkl:zstd. It's based on tikv master from a couple of days ago.)  

Starting out with an empty cluster and inserting using `raw_batch_put`s a panic occurs after a short period of time:
```
rocksdb background error. db: kv, reason: compaction, error: Corruption: Out-of-order key insertion into table /data/db/000057.sst
```
I have tried several different configuration variations and the common denominator for when it crashes seems to be the titan dictionary compression. Dictionary compression in normal sst files seems to be working fine. It seems to trigger more often/quicker when inserting with higher concurrency.

Any ideas on what might be going on here?

Setup:
- 3 tikv nodes (running inside docker containers on separate servers)
- max-replicas: 1
- avg blob value size ~200-300kB (can go up to 5MB)
- inserting up to 5000 key value pairs per second

Some sample titan configs:
```
min-blob-size = "8KB"
blob-file-compression = "zstd"
blob-file-compression-level = 3
blob-file-zstd-compression-dict-size = 1048576 # 1MB
blob-file-zstd-compression-sample-size = 16777216 # 16 MB
```

The full error message on one tikv node:
```
[2021/08/03 11:22:05.417 +00:00] [INFO] [raft.rs:2589] ["restored snapshot"] [snapshot_term=10] [snapshot_index=1574] [last_term=10] [last_index=1574] [commit=1574] [raft_id=2057] [region_id=989]
[2021/08/03 11:22:05.417 +00:00] [INFO] [raft.rs:2470] ["[commit: 1574, term: 10] restored snapshot [index: 1574, term: 10]"] [snapshot_term=10] [snapshot_index=1574] [commit=1574] [term=10] [raft_id=2057] [region_id=989]
[2021/08/03 11:22:05.417 +00:00] [INFO] [peer_storage.rs:1262] ["begin to apply snapshot"] [peer_id=2057] [region_id=989]
[2021/08/03 11:22:05.417 +00:00] [INFO] [peer_storage.rs:1305] ["apply snapshot with state ok"] [state="applied_index: 1574 truncated_state { index: 1574 term: 10 }"] [region="id: 989 start_key: 72752E65646162657A2F7461672F62657A2D6B6172746F73686B693A72753A7275 end_key: 72752E67616D656D61672E666F72756D732F746F7069632F3133323233372D63616C6C2D6F662D647574792D6D6F6465726E2D776172666172652D323031392F3F706167653D3132393A656E3A7573 region_epoch { conf_ver: 10 version: 11 } peers { id: 1355 store_id: 1 } peers { id: 2057 store_id: 4 role: Learner }"] [peer_id=2057] [region_id=989]
[2021/08/03 11:22:05.418 +00:00] [INFO] [peer.rs:3054] ["snapshot is applied"] [region="id: 989 start_key: 72752E65646162657A2F7461672F62657A2D6B6172746F73686B693A72753A7275 end_key: 72752E67616D656D61672E666F72756D732F746F7069632F3133323233372D63616C6C2D6F662D647574792D6D6F6465726E2D776172666172652D323031392F3F706167653D3132393A656E3A7573 region_epoch { conf_ver: 10 version: 11 } peers { id: 1355 store_id: 1 } peers { id: 2057 store_id: 4 role: Learner }"] [peer_id=2057] [region_id=989]
[2021/08/03 11:22:07.395 +00:00] [INFO] [size.rs:168] ["approximate size over threshold, need to do split check"] [threshold=1509949440] [size=3329110344] [region_id=2070]
[2021/08/03 11:22:07.450 +00:00] [INFO] [split_check.rs:289] ["update approximate size and keys with accurate value"] [keys=6787] [size=1006603116] [region_id=2070]
[2021/08/03 11:22:14.704 +00:00] [FATAL] [lib.rs:463] ["rocksdb background error. db: kv, reason: compaction, error: Corruption: Out-of-order key insertion into table /data/db/001582.sst"] [backtrace="stack backtrace:\n   0: tikv_util::set_panic_hook::{{closure}}\n             at components/tikv_util/src/lib.rs:462\n   1: std::panicking::rust_panic_with_hook\n             at library/std/src/panicking.rs:595\n   2: std::panicking::begin_panic_handler::{{closure}}\n             at library/std/src/panicking.rs:497\n   3: std::sys_common::backtrace::__rust_end_short_backtrace\n             at library/std/src/sys_common/backtrace.rs:141\n   4: rust_begin_unwind\n             at library/std/src/panicking.rs:493\n   5: std::panicking::begin_panic_fmt\n             at library/std/src/panicking.rs:435\n   6: <engine_rocks::event_listener::RocksEventListener as rocksdb::event_listener::EventListener>::on_background_error\n             at components/engine_rocks/src/event_listener.rs:105\n   7: rocksdb::event_listener::on_background_error\n             at root/.cargo/git/checkouts/rust-rocksdb-a9a28e74c6ead8ef/12a953b/src/event_listener.rs:340\n   8: _ZN24crocksdb_eventlistener_t17OnBackgroundErrorEN7rocksdb21BackgroundErrorReasonEPNS0_6StatusE\n             at crocksdb/c.cc:2340\n   9: _ZN7rocksdb12EventHelpers23NotifyOnBackgroundErrorERKSt6vectorISt10shared_ptrINS_13EventListenerEESaIS4_EENS_21BackgroundErrorReasonEPNS_6StatusEPNS_17InstrumentedMutexEPb\n             at root/.cargo/git/checkouts/rust-rocksdb-a9a28e74c6ead8ef/12a953b/librocksdb_sys/rocksdb/db/event_helpers.cc:53\n  10: _ZN7rocksdb12ErrorHandler10SetBGErrorERKNS_6StatusENS_21BackgroundErrorReasonE\n             at root/.cargo/git/checkouts/rust-rocksdb-a9a28e74c6ead8ef/12a953b/librocksdb_sys/rocksdb/db/error_han
dler.cc:219\n  11: _ZN7rocksdb6DBImpl20BackgroundCompactionEPbPNS_10JobContextEPNS_9LogBufferEPNS0_19PrepickedCompactionENS_3Env8PriorityE\n             at root/.cargo/git/checkouts/rust-rocksdb-a9a28e74c6ead8ef/12a953b/librocksdb_sys/rocksdb/db/db_impl/db_impl_compaction_flush.cc:2797\n  12: _ZN7rocksdb6DBImpl24BackgroundCallCompactionEPNS0_19PrepickedCompactionENS_3Env8PriorityE\n             at root/.cargo/git/checkouts/rust-rocksdb-a9a28e74c6ead8ef/12a953b/librocksdb_sys/rocksdb/db/db_impl/db_impl_compaction_flush.cc:2317\n  13: _ZN7rocksdb6DBImpl16BGWorkCompactionEPv\n             at root/.cargo/git/checkouts/rust-rocksdb-a9a28e74c6ead8ef/12a953b/librocksdb_sys/rocksdb/db/db_impl/db_impl_compaction_flush.cc:2091\n  14: _ZNKSt8functionIFvvEEclEv\n             at opt/rh/devtoolset-8/root/usr/include/c++/8/bits/std_function.h:687\n      _ZN7rocksdb14ThreadPoolImpl4Impl8BGThreadEm\n             at root/.cargo/git/checkouts/rust-rocksdb-a9a28e74c6ead8ef/12a953b/librocksdb_sys/rocksdb/util/threadpool_imp.cc:266\n  15: _ZN7rocksdb14ThreadPoolImpl4Impl15BGThreadWrapperEPv\n             at root/.cargo/git/checkouts/rust-rocksdb-a9a28e74c6ead8ef/12a953b/librocksdb_sys/rocksdb/util/threadpool_imp.cc:307\n  16: execute_native_thread_routine\n  17: <unknown>\n  18: clone\n"] [location=components/engine_rocks/src/event_listener.rs:105] [thread_name=<unnamed>]
```
and an excerpt from the `rocksdb.info` log file:
```
[2021/08/03 16:35:19.218 +00:00][2][INFO] [db/db_impl/db_impl_write.cc:1817] [default] New memtable created with log file: #468. Immutable memtables: 0.
[2021/08/03 16:35:19.218 +00:00][2][INFO] [db/column_family.cc:843] [default] Increasing compaction threads because we have 19 level-0 files 
[2021/08/03 16:35:19.218 +00:00][2][INFO] (Original Log Time 2021/08/03-16:35:19.218074) [db/db_impl/db_impl_compaction_flush.cc:2597] Compaction nothing to do
[2021/08/03 16:35:19.218 +00:00][2][INFO] [db/db_impl/db_impl_compaction_flush.cc:105] [JOB 47] Syncing log #440
[2021/08/03 16:35:19.218 +00:00][2][INFO] (Original Log Time 2021/08/03-16:35:19.218097) [db/db_impl/db_impl_compaction_flush.cc:2191] Calling FlushMemTableToOutputFile with column family [default], flush slots available 4, compaction slots available 6, flush slots scheduled 4, compaction slots scheduled 2
[2021/08/03 16:35:19.218 +00:00][2][INFO] [db/flush_job.cc:320] [default] [JOB 47] Flushing memtable with next log file: 468
[2021/08/03 16:35:19.218 +00:00][2][INFO] EVENT_LOG_v1 {"time_micros": 1628008519218305, "job": 47, "event": "flush_started", "num_memtables": 1, "num_entries": 44124, "num_deletes": 0, "total_data_size": 57895565, "memory_usage": 59209888, "flush_reason": "Manual Flush"}
[2021/08/03 16:35:19.218 +00:00][2][INFO] [db/flush_job.cc:349] [default] [JOB 47] Level-0 flush table #469: started
[2021/08/03 16:35:19.218 +00:00][2][INFO] [able_builder.cc:157] Titan table builder created new blob file 514.
[2021/08/03 16:35:20.043 +00:00][2][ERROR] [table/block_based/block_based_table_builder.cc:507] Out-of-order key insertion into block based table /data/db/000469.sst
[2021/08/03 16:35:20.053 +00:00][2][WARN] [able_builder.cc:246] Titan table builder finish failed. Delete output file 514.
[2021/08/03 16:35:20.054 +00:00][2][INFO] [db/flush_job.cc:390] [default] [JOB 47] Level-0 flush table #469: 0 bytes Corruption: Out-of-order key insertion into table /data/db/000469.sst
```

The sst file number differs because it is from another run where titan/tikv panicked.

I am happy to provide more information.

**Comment by Connor1996** [comment]
Titan zstd dict compression is not GA yet. It should be a bug of titan zstd compression. Would you mind giving us your write workload?

**Comment by Connor1996** [comment]
I think it's probably due to #232

**Comment by Connor1996** [comment]
Fixed by #232
